New structure for contributing docs

dnephin opened the pull request:
[Preview the docs](https://github.com/hashicorp/consul/tree/docs-pairing/contributing#readme).

This PR setups up a new structure for the docs under `contributing/`. These developer docs designed to be a high level guide to the codebase. In some cases we may have more specific diagrams (state, erd, sequence, flow, etc) for areas of the code that are particularly complex or are under active development.

There is still lots of work to do on these docs, but I think there is enough here to merge and allow more people to contribute in parallel.

This PR:
1. organizes the docs into categories. This categorization can evolve over time, but I think the current categories are a good compromise between existing code structure and ideal structure.
2. adds more detailed docs for "CLI" and "Cluster persistence", and starts some of the detailed docs for "RPC", and "ACL"
3. adds some notes to other sections to act as a guide for what docs to add to each section

I think the eventual goal will be to remove the `INTERNALS.md` file. I've left it for now, and removed a few of the sections that have been replaced. I think we are pretty close to being able to remove the file entirely.

jkirschner-hashicorp commented on contributing/README.md:
```diff
@@ -5,10 +5,33 @@ See [our contributing guide](../.github/CONTRIBUTING.md) to get started.
+
+[user documentation]: https://www.consul.io/docs
+
+![Overview](./overview.svg)
```
@dnephin: There's a typo in this image: `Service Mesh` is spelled as `Sercice Mesh`

dnephin replied:
Good spot! I pushed a fix

eculver commented on contributing/consul-mermaid-theme.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Is this meant to be included in the PR?

dnephin replied:
Ah yes, but I should probably include some details about it in the README.

This can be added as the config to https://mermaid-js.github.io/mermaid-live-editor/edit to allow anyone to make changes to the diagrams and keep the same style/colors. I'll add a note about the diagrams.

eculver commented on contributing/rpc/README.md:
```diff
@@ -0,0 +1,45 @@
+
+
+## RPC connections and load balancing
+
```
These are still a WIP right? Would it make sense to add a blurb that says: 
> This section is a work in progress, it will eventually cover topics like: ...

To be a little more clear as to what's going on here.

eculver reviewed (approved):
This is solid. As someone new to the codebase, it makes the docs seem much more "robust" and match the code more clearly. Nice work. Just a couple small comments.

dnephin commented:
Thanks for the review! I added a section in the README that links to the diagram editor and the theme.json. Also added the "this section is a work in progress" suggestion.

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/405470.
